### ISSUE TITLE
Is it normal for l_d_real and l_d_fake keep 0 during the first several epochs

### ISSUE THREAD

**Issue by Darrenxc** [Maclory/SPSR#8]
the val image results show black for all pixels.
I followed your code, and parameters are set by default.

20-04-26 15:14:21.084 - INFO: <epoch:  4, iter:   4,000> psnr: 5.9762e+00
20-04-26 15:16:38.846 - INFO: <epoch:  4, iter:   4,100, lr:1.000e-04> l_g_pix: 4.7353e-03 l_g_fea: 1.9221e+00 l_g_gan: 1.0035e-01 l_g_pix_grad_branch: 3.4439e-02 l_d_real: 0.0000e+00 l_d_fake: 0.0000e+00 l_d_real_grad: 0.0000e+00 l_d_fake_grad: 0.0000e+00 D_real: 2.8926e+01 D_fake: 8.8556e+00 D_real_grad: 3.2620e+01 D_fake_grad: -8.0358e+00
20-04-26 15:18:57.880 - INFO: <epoch:  4, iter:   4,200, lr:1.000e-04> l_g_pix: 4.5557e-03 l_g_fea: 2.1639e+00 l_g_gan: 7.4714e-02 l_g_pix_grad_branch: 5.0381e-02 l_d_real: 1.7411e-06 l_d_fake: 1.0290e-06 l_d_real_grad: 0.0000e+00 l_d_fake_grad: 0.0000e+00 D_real: 2.9009e+01 D_fake: 1.4066e+01 D_real_grad: 2.9313e+01 D_fake_grad: -7.0066e+00
20-04-26 15:21:17.073 - INFO: <epoch:  5, iter:   4,300, lr:1.000e-04> l_g_pix: 4.5929e-03 l_g_fea: 2.1910e+00 l_g_gan: 8.5829e-02 l_g_pix_grad_branch: 4.4242e-02 l_d_real: 1.3489e-07 l_d_fake: 2.9489e-07 l_d_real_grad: 0.0000e+00 l_d_fake_grad: 0.0000e+00 D_real: 2.9877e+01 D_fake: 1.2711e+01 D_real_grad: 3.1373e+01 D_fake_grad: -4.3015e+00
20-04-26 15:23:38.389 - INFO: <epoch:  5, iter:   4,400, lr:1.000e-04> l_g_pix: 4.3075e-03 l_g_fea: 2.0604e+00 l_g_gan: 1.2741e-01 l_g_pix_grad_branch: 4.0523e-02 l_d_real: 0.0000e+00 l_d_fake: 0.0000e+00 l_d_real_grad: 0.0000e+00 l_d_fake_grad: 0.0000e+00 D_real: 2.5259e+01 D_fake: -2.2342e-01 D_real_grad: 3.0326e+01 D_fake_grad: -3.8803e+00
20-04-26 15:25:56.332 - INFO: <epoch:  5, iter:   4,500, lr:1.000e-04> l_g_pix: 4.4438e-03 l_g_fea: 1.9418e+00 l_g_gan: 1.1235e-01 l_g_pix_grad_branch: 4.0387e-02 l_d_real: 0.0000e+00 l_d_fake: 0.0000e+00 l_d_real_grad: 0.0000e+00 l_d_fake_grad: 0.0000e+00 D_real: 2.7480e+01 D_fake: 5.0108e+00 D_real_grad: 3.3779e+01 D_fake_grad: -5.4768e+00
20-04-26 15:26:01.084 - INFO: # Validation # PSNR: 5.8998e+00
20-04-26 15:26:01.094 - INFO: <epoch:  5, iter:   4,500> psnr: 5.8998e+00
20-04-26 15:28:28.558 - INFO: <epoch:  5, iter:   4,600, lr:1.000e-04> l_g_pix: 4.0463e-03 l_g_fea: 2.3505e+00 l_g_gan: 7.1082e-02 l_g_pix_grad_branch: 4.0911e-02 l_d_real: 8.7050e-06 l_d_fake: 5.9227e-06 l_d_real_grad: 0.0000e+00 l_d_fake_grad: 0.0000e+00 D_real: 2.5983e+01 D_fake: 1.1766e+01 D_real_grad: 3.0323e+01 D_fake_grad: -7.1722e+00 

thank you

**Comment by LiquidAmmonia** [comment]
Here is our results in the similiar epoch:

20-03-11 03:11:43.329 - INFO: <epoch:  4, iter:   8,600, lr:1.000e-04> l_g_pix: 2.9860e-04 l_g_fea: 9.2371e-01 l_g_gan: 1.1278e-02 l_g_pix_grad_branch: 2.1001e-02 l_d_real: 1.1792e-01 l_d_fake: 1.1694e-01 l_d_real_grad: 1.3796e-01 l_d_fake_grad: 1.2376e-01 D_real: -4.8384e+01 D_fake: -5.0522e+01 D_real_grad: 3.5356e+00 D_fake_grad: 1.3205e+00 
20-03-11 03:13:41.126 - INFO: <epoch:  4, iter:   8,700, lr:1.000e-04> l_g_pix: 4.8472e-04 l_g_fea: 1.0007e+00 l_g_gan: 1.2774e-02 l_g_pix_grad_branch: 3.0137e-02 l_d_real: 1.4683e-01 l_d_fake: 1.2315e-01 l_d_real_grad: 8.4051e-01 l_d_fake_grad: 8.3112e-01 D_real: -2.6252e+01 D_fake: -2.8672e+01 D_real_grad: 4.6735e+00 D_fake_grad: 4.8747e+00 
20-03-11 03:15:38.536 - INFO: <epoch:  4, iter:   8,800, lr:1.000e-04> l_g_pix: 3.1398e-04 l_g_fea: 9.4902e-01 l_g_gan: 9.7581e-03 l_g_pix_grad_branch: 2.2440e-02 l_d_real: 1.6637e-01 l_d_fake: 1.6640e-01 l_d_real_grad: 5.2797e-02 l_d_fake_grad: 4.4843e-02 D_real: -1.5506e+01 D_fake: -1.7291e+01 D_real_grad: -1.4672e+01 D_fake_grad: -1.7963e+01 
20-03-11 03:17:33.011 - INFO: <epoch:  4, iter:   8,900, lr:1.000e-04> l_g_pix: 4.9204e-04 l_g_fea: 1.1781e+00 l_g_gan: 9.9414e-03 l_g_pix_grad_branch: 3.5996e-02 l_d_real: 1.4888e-01 l_d_fake: 1.4886e-01 l_d_real_grad: 3.7366e-01 l_d_fake_grad: 3.7220e-01 D_real: -2.6956e+01 D_fake: -2.8795e+01 D_real_grad: -3.5206e+00 D_fake_grad: -4.3362e+00 
20-03-11 03:19:27.311 - INFO: <epoch:  4, iter:   9,000, lr:1.000e-04> l_g_pix: 5.1673e-04 l_g_fea: 1.2416e+00 l_g_gan: 5.9759e-03 l_g_pix_grad_branch: 3.4374e-02 l_d_real: 4.0131e-01 l_d_fake: 4.0127e-01 l_d_real_grad: 1.4362e-04 l_d_fake_grad: 1.5229e-04 D_real: -3.2262e+01 D_fake: -3.3056e+01 D_real_grad: 1.3240e+01 D_fake_grad: 4.1770e+00 
20-03-11 03:21:21.307 - INFO: <epoch:  4, iter:   9,100, lr:1.000e-04> l_g_pix: 3.4264e-04 l_g_fea: 9.9129e-01 l_g_gan: 8.9233e-03 l_g_pix_grad_branch: 2.7327e-02 l_d_real: 2.2853e-01 l_d_fake: 2.2768e-01 l_d_real_grad: 3.1390e-03 l_d_fake_grad: 2.7141e-03 D_real: -1.7226e+01 D_fake: -1.8782e+01 D_real_grad: -8.1139e+00 D_fake_grad: -1.4139e+01 
20-03-11 03:23:17.438 - INFO: <epoch:  4, iter:   9,200, lr:1.000e-04> l_g_pix: 3.2942e-04 l_g_fea: 1.0880e+00 l_g_gan: 1.2743e-02 l_g_pix_grad_branch: 2.4977e-02 l_d_real: 9.6658e-02 l_d_fake: 1.0251e-01 l_d_real_grad: 4.9492e-02 l_d_fake_grad: 4.9627e-02 D_real: -2.9218e+00 D_fake: -5.3708e+00 D_real_grad: 1.6236e+01 D_fake_grad: 1.2494e+01 
20-03-11 03:25:14.717 - INFO: <epoch:  4, iter:   9,300, lr:1.000e-04> l_g_pix: 4.7350e-04 l_g_fea: 1.1153e+00 l_g_gan: 2.1990e-02 l_g_pix_grad_branch: 3.0954e-02 l_d_real: 1.4940e-02 l_d_fake: 1.4847e-02 l_d_real_grad: 1.1710e-02 l_d_fake_grad: 1.2022e-02 D_real: -4.4433e+01 D_fake: -4.8816e+01 D_real_grad: -2.0744e+01 D_fake_grad: -2.5300e+01 
20-03-11 03:27:11.598 - INFO: <epoch:  4, iter:   9,400, lr:1.000e-04> l_g_pix: 3.3018e-04 l_g_fea: 8.2489e-01 l_g_gan: 1.8528e-03 l_g_pix_grad_branch: 2.3327e-02 l_d_real: 1.1791e+00 l_d_fake: 1.1803e+00 l_d_real_grad: 1.0163e+00 l_d_fake_grad: 1.0258e+00 D_real: -6.8521e+00 D_fake: -6.0429e+00 D_real_grad: -1.6011e+01 D_fake_grad: -1.5475e+01 
20-03-11 03:29:10.859 - INFO: <epoch:  4, iter:   9,500, lr:1.000e-04> l_g_pix: 4.3919e-04 l_g_fea: 1.0775e+00 l_g_gan: 1.5230e-02 l_g_pix_grad_branch: 3.1943e-02 l_d_real: 5.1752e-02 l_d_fake: 5.2153e-02 l_d_real_grad: 3.4912e-01 l_d_fake_grad: 3.4781e-01 D_real: -6.5865e+01 D_fake: -6.8859e+01 D_real_grad: -1.6757e+00 D_fake_grad: -2.6389e+00 
20-03-11 03:31:08.209 - INFO: <epoch:  4, iter:   9,600, lr:1.000e-04> l_g_pix: 6.0910e-04 l_g_fea: 1.3984e+00 l_g_gan: 2.6751e-02 l_g_pix_grad_branch: 3.9058e-02 l_d_real: 5.0800e-03 l_d_fake: 4.9483e-03 l_d_real_grad: 1.4255e+00 l_d_fake_grad: 1.4420e+00 D_real: -2.4880e+01 D_fake: -3.0226e+01 D_real_grad: 2.4057e+01 D_fake_grad: 2.5116e+01 
20-03-11 03:33:05.460 - INFO: <epoch:  4, iter:   9,700, lr:1.000e-04> l_g_pix: 3.2834e-04 l_g_fea: 1.1017e+00 l_g_gan: 1.7542e-02 l_g_pix_grad_branch: 2.8164e-02 l_d_real: 3.4227e-02 l_d_fake: 3.4031e-02 l_d_real_grad: 2.1086e-01 l_d_fake_grad: 2.2272e-01 D_real: 3.5049e+01 D_fake: 3.1574e+01 D_real_grad: -4.5340e+00 D_fake_grad: -6.0704e+00 
20-03-11 03:35:03.667 - INFO: <epoch:  4, iter:   9,800, lr:1.000e-04> l_g_pix: 6.0577e-04 l_g_fea: 1.1519e+00 l_g_gan: 1.7393e-02 l_g_pix_grad_branch: 3.8009e-02 l_d_real: 4.2432e-02 l_d_fake: 3.8658e-02 l_d_real_grad: 3.7895e-03 l_d_fake_grad: 3.7555e-03 D_real: -5.6270e+01 D_fake: -5.9708e+01 D_real_grad: 6.3120e+00 D_fake_grad: 6.5626e-01 
20-03-11 03:37:00.335 - INFO: <epoch:  4, iter:   9,900, lr:1.000e-04> l_g_pix: 4.0234e-04 l_g_fea: 9.9020e-01 l_g_gan: 3.8569e-03 l_g_pix_grad_branch: 2.7882e-02 l_d_real: 6.3786e-01 l_d_fake: 6.3671e-01 l_d_real_grad: 1.8050e-02 l_d_fake_grad: 1.5590e-02 D_real: 2.3528e+01 D_fake: 2.3394e+01 D_real_grad: -1.2171e+00 D_fake_grad: -5.4544e+00 
20-03-11 03:38:57.460 - INFO: <epoch:  4, iter:  10,000, lr:1.000e-04> l_g_pix: 3.7156e-04 l_g_fea: 9.4680e-01 l_g_gan: 3.7199e-02 l_g_pix_grad_branch: 2.9080e-02 l_d_real: 6.0010e-04 l_d_fake: 5.9699e-04 l_d_real_grad: 2.6952e-02 l_d_fake_grad: 2.4174e-02 D_real: 2.1714e+01 D_fake: 1.4274e+01 D_real_grad: 4.2452e+00 D_fake_grad: 4.5549e-01 
20-03-11 03:39:13.748 - INFO: # Validation # PSNR: 2.4245e+01
20-03-11 03:39:13.759 - INFO: <epoch:  4, iter:  10,000> psnr: 2.4245e+01

 My suggestion is to check your configurations and paths again. If something is still wrong, please provide more infomation so we could help you more.

**Comment by Darrenxc** [comment]
val images become normal (showing edges) after the first training state sample interval.
However, I still don't know why these values keep zeros at first, and suddenly becomes normal. 
here comes my training process this morning. 
20-04-27 08:16:28.127 - INFO: # Validation # PSNR: 2.3405e+01
20-04-27 08:16:28.138 - INFO: <epoch: 37, iter:  32,000> psnr: 2.3405e+01
20-04-27 08:20:07.969 - INFO: <epoch: 37, iter:  32,100, lr:1.000e-04> l_g_pix: 3.9904e-04 l_g_fea: 1.1266e+00 l_g_gan: 1.5145e-02 l_g_pix_grad_branch: 2.1599e-02 l_d_real: 5.3987e-02 l_d_fake: 5.3480e-02 l_d_real_grad: 6.9928e-03 l_d_fake_grad: 7.6716e-03 D_real: -2.0411e+00 D_fake: -5.0164e+00 D_real_grad: -2.1960e+01 D_fake_grad: -2.7222e+01
20-04-27 08:23:49.254 - INFO: <epoch: 38, iter:  32,200, lr:1.000e-04> l_g_pix: 4.3509e-04 l_g_fea: 1.1062e+00 l_g_gan: 8.5414e-03 l_g_pix_grad_branch: 2.4852e-02 l_d_real: 2.3740e-01 l_d_fake: 2.2797e-01 l_d_real_grad: 4.3040e-06 l_d_fake_grad: 1.2893e-06 D_real: 3.8903e+01 D_fake: 3.7428e+01 D_real_grad: 1.7006e+00 D_fake_grad: -1.2137e+01
20-04-27 08:27:29.444 - INFO: <epoch: 38, iter:  32,300, lr:1.000e-04> l_g_pix: 4.4393e-04 l_g_fea: 1.1402e+00 l_g_gan: 2.0608e-03 l_g_pix_grad_branch: 2.4088e-02 l_d_real: 1.3242e+00 l_d_fake: 1.3502e+00 l_d_real_grad: 2.4406e-06 l_d_fake_grad: 2.2869e-06 D_real: -3.8031e+01 D_fake: -3.7106e+01 D_real_grad: -4.9506e+01 D_fake_grad: -6.2692e+01
20-04-27 08:31:09.819 - INFO: <epoch: 38, iter:  32,400, lr:1.000e-04> l_g_pix: 4.9109e-04 l_g_fea: 1.2541e+00 l_g_gan: 3.7171e-02 l_g_pix_grad_branch: 2.5688e-02 l_d_real: 6.3716e-04 l_d_fake: 6.6270e-04 l_d_real_grad: 1.7814e-02 l_d_fake_grad: 1.8007e-02 D_real: 3.7315e+01 D_fake: 2.9881e+01 D_real_grad: -5.0757e+01 D_fake_grad: -5.5080e+01
20-04-27 08:34:51.019 - INFO: <epoch: 38, iter:  32,500, lr:1.000e-04> l_g_pix: 4.1007e-04 l_g_fea: 1.1735e+00 l_g_gan: 1.0017e-02 l_g_pix_grad_branch: 2.3071e-02 l_d_real: 1.6023e-01 l_d_fake: 1.6361e-01 l_d_real_grad: 1.0146e-02 l_d_fake_grad: 1.1829e-02 D_real: 1.1412e+01 D_fake: 9.5708e+00 D_real_grad: -4.9027e+01 D_fake_grad: -5.3763e+01
20-04-27 08:34:58.123 - INFO: # Validation # PSNR: 2.3439e+01
20-04-27 08:34:58.128 - INFO: <epoch: 38, iter:  32,500> psnr: 2.3439e+01 
![baboonx4_32000](https://user-images.githubusercontent.com/57433102/80324339-0e395680-8863-11ea-847f-fed9127ff0cf.png)

my setting parameters showing below:

{
  "name": "SPSR"   
  , "use_tb_logger": true
  , "model":"spsr"
  , "scale": 4
  , "gpu_ids": [1]

  , "datasets": {
    "train": {
      "name": "DIV2K"
      , "mode": "LRHR"
      , "dataroot_HR": "../../datasets/DIV2K/DIV2K800_train_sub.lmdb"
      , "dataroot_LR": "../../datasets/DIV2K/DIV2K800_LRtrain_sub.lmdb"
      , "subset_file": null
      , "use_shuffle": true
      , "n_workers": 16
      , "batch_size": 38
      , "HR_size": 128
      , "use_flip": true
      , "use_rot": true
    }
    , "val": {
      "name": "Set14"
      , "mode": "LRHR"
      , "dataroot_HR": "../../datasets/benchmark/Set14/HR"
      , "dataroot_LR": "../../datasets/benchmark/Set14/LR_bicubic/X4"
    }
  }

  , "path": {
    "root": "../training_res/" //modify to you own root path
    // , "resume_state": "../experiments/SPSR_0311/training_state/25000.state"
    , "pretrain_model_G": "../experiments/pretrained_models/RRDB_PSNR_x4.pth"
  }

  , "network_G": {
    "which_model_G": "spsr_net" 
    , "norm_type": null
    , "mode": "CNA"
    , "nf": 64
    , "nb": 23
    , "in_nc": 3
    , "out_nc": 3
    , "gc": 32
    , "group": 1
  }

  , "network_D": {
    "which_model_D": "discriminator_vgg_128"
    , "norm_type": "batch"
    , "act_type": "leakyrelu"
    , "mode": "CNA"
    , "nf": 64
    , "in_nc": 3
  }

  , "train": {
    "lr_G": 1e-4
    , "lr_G_grad": 1e-4
    , "weight_decay_G": 0
    , "weight_decay_G_grad": 0
    , "beta1_G": 0.9
    , "beta1_G_grad": 0.9
    , "lr_D": 1e-4
    , "weight_decay_D": 0
    , "beta1_D": 0.9
    , "lr_scheme": "MultiStepLR"
    , "lr_steps": [50000, 100000, 200000, 300000]
    , "lr_gamma": 0.5

    , "pixel_criterion": "l1"
    , "pixel_weight": 1e-2
    , "feature_criterion": "l1"
    , "feature_weight": 1
    , "gan_type": "vanilla"
    , "gan_weight": 5e-3
    , "gradient_pixel_weight": 1e-2
    , "gradient_gan_weight": 5e-3
    , "pixel_branch_criterion": "l1"
    , "pixel_branch_weight": 5e-1
    , "Branch_pretrain" : 1
    , "Branch_init_iters" : 5000

    , "manual_seed": 9
    , "niter": 5e6
    , "val_freq": 500
  }

  , "logger": {
    "print_freq": 100
    , "save_checkpoint_freq": 5e3
  }
}

Thank you for your amazing work

**Comment by LiquidAmmonia** [comment]
In our early experiments, we found that training directly will result in unstable gradient results, because the parameters of the main branch have been pre-trained but the parameters of the gradient branch haven't. Therefore, we first fixed all the parameters in the main branch to pre-train the gradient branch. This process is controlled by hyperparameter 'Branch_init_iters'(5000 by default) in the configurations. Before 'Branch_init_iters' just the parameters in the gradient branch are optimized. This may explain the phenomenon that the initial output image is black.
